src/messaging/ExchangeContext.h

```diff
@@ -0,0 +1,271 @@
+     *                           the exchange.
+     */
+    void SetInitiator(bool inInitiator);
+
+    /**
+     *  Set whether a response is expected on this exchange.
+     *
+     *  @param[in]  inResponseExpected  A Boolean indicating whether (true) or not
+     *                                  (false) a response is expected on this
+     *                                  exchange.
+     */
+    void SetResponseExpected(bool inResponseExpected);
+
+    /**
+     *  Send a CHIP message on this exchange.
+     *
+     *  @param[in]    protocolId     The profile identifier of the CHIP message to be sent.
+     *
+     *  @param[in]    msgType       The message type of the corresponding profile.
+     *
+         *  @param[in]    msgPayload    A pointer to the PacketBuffer object holding the CHIP message.
+     *
+     *  @param[in]    sendFlags     Flags set by the application for the CHIP message being sent.
+     *
+     *  @param[in]    msgCtxt       A pointer to an application-specific context object to be associated
+     *                              with the message being sent.
+
+     *  @retval  #CHIP_ERROR_INVALID_ARGUMENT               if an invalid argument was passed to this SendMessage API.
+     *  @retval  #CHIP_ERROR_SEND_THROTTLED                 if this exchange context has been throttled when using the
+     *                                                       CHIP reliable messaging protocol.
+     *  @retval  #CHIP_ERROR_WRONG_MSG_VERSION_FOR_EXCHANGE if there is a mismatch in the specific send operation and the
+     *                                                       CHIP message protocol version that is supported. For example,
+     *                                                       this error would be generated if CHIP Reliable Messaging
+     *                                                       semantics are being attempted when the CHIP message protocol
+     *                                                       version is V1.
+     *  @retval  #CHIP_ERROR_NOT_CONNECTED                  if the context was associated with a connection that is now
+     *                                                       closed.
+     *  @retval  #CHIP_ERROR_INCORRECT_STATE                if the state of the exchange context is incorrect.
+     *  @retval  #CHIP_NO_ERROR                             if the CHIP layer successfully sent the message down to the
+     *                                                       network layer.
+     */
+    CHIP_ERROR SendMessage(uint32_t protocolId, uint8_t msgType, System::PacketBuffer * msgPayload, uint16_t sendFlags = 0,
+                           void * msgCtxt = nullptr);
+
+    /**
+     *  Tear down the Trickle retransmission mechanism by canceling the periodic timers
+     *  within Trickle and freeing the message buffer holding the Weave
+     *  message.
+     */
+    void TeardownTrickleRetransmit();
```

Comment:
could we separate trickle from the first PR? It seems that CRMP and trickling are additional functionality

Reply:
CRMP is message layer internal staff, this is protocol level feature (Trickle rebroadcast algorithm's periodic retransmission timer mechanism.)

Reply:
This feature is not part of CRMP, but it is also not discussed within the committee, I am checking if we should remove this feature from CHIP v1.0

Reply:
Trickle retransmission is another mechanism of re-transmission, this feature can not be used together with CRMP. Since I don't see any agreed protocols plan to use this feature, clean it up now.